--- a/android_configure.py
+++ b/android_configure.py
@@ -55,7 +55,7 @@
 print("\033[92mInfo: \033[0m" + "Configuring for " + DEST_CPU + "...")

 if platform.system() == "Darwin":
-    host_os = "darwin"
+    host_os = "mac"
     toolchain_path = android_ndk_path + "/toolchains/llvm/prebuilt/darwin-x86_64"

 elif platform.system() == "Linux":
@@ -71,6 +71,8 @@
 GYP_DEFINES += " android_target_arch=" + arch
 GYP_DEFINES += " host_os=" + host_os + " OS=android"
 GYP_DEFINES += " android_ndk_path=" + android_ndk_path
+if os.environ.get("NODE_ANDROID_HOST_ARCH"):
+    GYP_DEFINES += " host_arch=" + os.environ["NODE_ANDROID_HOST_ARCH"]
 os.environ['GYP_DEFINES'] = GYP_DEFINES

 if os.path.exists("./configure"):
--- a/deps/uv/uv.gyp
+++ b/deps/uv/uv.gyp
@@ -270,7 +270,7 @@
             '-fno-strict-aliasing',
           ],
         }],
-        [ 'OS in "mac ios"', {
+        [ 'OS in "mac ios" or (_toolset=="host" and host_os=="mac")', {
           'sources': [
             '<@(uv_sources_apple)',
           ],
@@ -288,7 +288,7 @@
             'libraries': [ '-ldl', '-lrt' ],
           },
         }],
-        [ 'OS=="android"', {
+        [ 'OS=="android" and _toolset=="target"', {
           'defines': [
             '_GNU_SOURCE',
           ],
@@ -392,7 +392,7 @@
             'src/unix/bsd-proctitle.c'
           ],
         }],
-        [ 'OS in "ios mac freebsd dragonflybsd openbsd netbsd".split()', {
+        [ 'OS in "ios mac freebsd dragonflybsd openbsd netbsd".split() or (_toolset=="host" and host_os=="mac")', {
           'sources': [
             '<@(uv_sources_bsd_common)',
           ],
--- a/deps/zlib/zlib.gyp
+++ b/deps/zlib/zlib.gyp
@@ -186,6 +186,11 @@
             'include_dirs': [ '<(ZLIB_ROOT)' ],
           },
           'conditions': [
+            ['OS=="android"', {
+              'sources': [
+                '<(ZLIB_ROOT)/android-cpu-features.c',
+              ],
+            }],
             ['OS!="win"', {
               'cflags!': [ '-ansi' ],
               'cflags': [ '-Wno-implicit-fallthrough' ],
--- a/tools/v8_gypfiles/v8.gyp
+++ b/tools/v8_gypfiles/v8.gyp
@@ -1176,6 +1176,12 @@
                     '<(V8_ROOT)/src/trap-handler/handler-outside-posix.cc',
                   ],
                 }],
+                ['_toolset=="host" and host_os=="mac" and OS=="android"', {
+                  'sources': [
+                    '<(V8_ROOT)/src/trap-handler/handler-inside-posix.cc',
+                    '<(V8_ROOT)/src/trap-handler/handler-outside-posix.cc',
+                  ],
+                }],
                 ['OS=="win"', {
                   'sources': [
                     '<(V8_ROOT)/src/trap-handler/handler-inside-win.cc',
@@ -1204,6 +1210,12 @@
                     '<(V8_ROOT)/src/trap-handler/handler-outside-posix.cc',
                   ],
                 }],
+                ['_toolset=="host" and host_os=="mac" and OS=="android"', {
+                  'sources': [
+                    '<(V8_ROOT)/src/trap-handler/handler-inside-posix.cc',
+                    '<(V8_ROOT)/src/trap-handler/handler-outside-posix.cc',
+                  ],
+                }],
                 ['(_toolset=="host" and host_arch=="x64" or _toolset=="target" and target_arch=="x64") and OS=="win"', {
                   'sources': [
                     '<(V8_ROOT)/src/trap-handler/handler-inside-win.cc',
@@ -1502,12 +1514,13 @@
             ],
           },
           'target_conditions': [
-            ['_toolset=="host"', {
+            ['_toolset=="host" and host_os=="linux"', {
               'sources': [
                 '<(V8_ROOT)/src/base/debug/stack_trace_posix.cc',
                 '<(V8_ROOT)/src/base/platform/platform-linux.cc',
               ],
-            }, {
+            }],
+            ['_toolset=="target"', {
               'sources': [
                 '<(V8_ROOT)/src/base/debug/stack_trace_android.cc',
                 '<(V8_ROOT)/src/base/platform/platform-linux.cc',
--- a/tools/gyp/pylib/gyp/generator/make.py
+++ b/tools/gyp/pylib/gyp/generator/make.py
@@ -225,7 +225,7 @@
 cmd_alink = rm -f $@ && $(AR.$(TOOLSET)) crs $@ $(filter %.o,$^)

 quiet_cmd_alink_thin = AR($(TOOLSET)) $@
-cmd_alink_thin = rm -f $@ && $(AR.$(TOOLSET)) crsT $@ $(filter %.o,$^)
+cmd_alink_thin = rm -f $@ && $(AR.$(TOOLSET)) crs$(if $(filter target,$(TOOLSET)),T,) $@ $(filter %.o,$^)

 # Note: this does not handle spaces in paths
 define xargs
@@ -249,7 +249,7 @@
 define create_thin_archive
         rm -f $(1) $(OBJ_FILE_LIST); mkdir -p `dirname $(1)`
         $(call write-to-file,$(1).$(OBJ_FILE_LIST),$(filter %.o,$(2)))
-        $(AR.$(TOOLSET)) crsT $(1) @$(1).$(OBJ_FILE_LIST)
+        $(AR.$(TOOLSET)) crs$(if $(filter target,$(TOOLSET)),T,) $(1) @$(1).$(OBJ_FILE_LIST)
 endef

 # Due to circular dependencies between libraries :(, we wrap the
@@ -258,7 +258,7 @@
 quiet_cmd_link = LINK($(TOOLSET)) $@
 quiet_cmd_link_host = LINK($(TOOLSET)) $@
 cmd_link = $(LINK.$(TOOLSET)) $(GYP_LDFLAGS) $(LDFLAGS.$(TOOLSET)) -o $@ -Wl,--start-group $(LD_INPUTS) -Wl,--end-group $(LIBS)
-cmd_link_host = $(LINK.$(TOOLSET)) $(GYP_LDFLAGS) $(LDFLAGS.$(TOOLSET)) -o $@ -Wl,--start-group $(LD_INPUTS) -Wl,--end-group $(LIBS)
+cmd_link_host = $(LINK.$(TOOLSET)) $(GYP_LDFLAGS) $(LDFLAGS.$(TOOLSET)) -o $@ $(LD_INPUTS) $(LIBS)

 # Other shared-object link notes:
 # - Set SONAME to the library filename so our binaries don't reference
